src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/Regex.cs

```diff
@@ -435,116 +458,88 @@ internal void RunAllMatchesWithCallback<TState>(string input, int startat, ref T
             RegexRunner runner = Interlocked.Exchange(ref _runner, null) ?? CreateRunner();
             try
             {
-                // For the string overload, we need to set runtext before starting the match attempts.
+                // We need to set runtext before starting the match attempts.
                 runner.runtext = input;
-                RunAllMatchesWithCallbackHelper(input, startat, ref state, callback, runner, usingStringOverload: true, reuseMatchObject);
```

Comment:
All of the following changes are just due to the fact that we removed RunAllMatchesWithCallback span overload since it is not needed any longer (as Count is implemented on top of Enumerate now) and we no longer need RunAllMatchesCallbackHelper to be a separate method since it was only there so the string and span overload could share code but the span one is gone now, so the following changes are just collapsing the helper inside the string overload and adjusting the comments.

Reply:
> as Count is implemented on top of Enumerate now

Can you share perf numbers for this?

Reply:
Sure. Not exactly the same, but since we didn't have benchmarks for Count(span) I had locally also modified Count(string) to work over enumerate and ran the Sherlock benchmarks and these were the results https://gist.github.com/joperezr/7709f4e45387d250b009f97ead72e3d5

That said I ended up undoing the change on Count(string) to use Enumerate as that broke the precompiled scenario. The results above showed that working over enumerate as opposed to what we used to do was only a very slight improvement but perf was almost on par.

I can get benchmarks for the Count(span) overload before and after in the next couple of days.

Reply:
Thanks. To understand worst case, can you try counting a long string of all a's with a pattern that's just a single a?

Reply:
Will do

Reply:
Is it worth adding a benchmark to dotnet/performance? Or, maybe any regression in this would almost surely show up in an existing one.

Reply:
It wouldn't show up today, as all of our benchmarks go against the string overload which I am not changing here. It will for sure be a good idea to add some benchmarks against the span overload in dotnet/performance. I have ran the benchmark that @stephentoub asked for above, and it does look like we are regressing Count(span) by making this change for that case, I suppose mainly cause before we were in a tighter loop as well as we wouldn't call Tidy on the match and would only return null when a match was found. Here are the results that I'm seeing locally:

``` ini

BenchmarkDotNet=v0.13.1, OS=Windows 10.0.22000
Intel Core i7-8700 CPU 3.20GHz (Coffee Lake), 1 CPU, 12 logical and 6 physical cores
.NET SDK=7.0.100-preview.2.22110.1
  [Host]     : .NET 7.0.0 (7.0.22.10302), X64 RyuJIT
  Job-OZRCZA : .NET 7.0.0 (42.42.42.42424), X64 RyuJIT
  Job-PJSYDW : .NET 7.0.0 (42.42.42.42424), X64 RyuJIT


```
| Method |        Job |         Toolchain |         Options | NumberOfAs |        Mean |       Error |      StdDev | Ratio | RatioSD | Allocated |
|------- |----------- |------------------ |---------------- |----------- |------------:|------------:|------------:|------:|--------:|----------:|
|  **Count** | **Job-OZRCZA** | **\main\corerun.exe** |            **None** |      **10000** |    **533.8 μs** |     **8.59 μs** |     **7.62 μs** |  **1.00** |    **0.00** |         **-** |
|  Count | Job-PJSYDW |   \pr\corerun.exe |            None |      10000 |    652.3 μs |    12.72 μs |    16.09 μs |  1.22 |    0.03 |         - |
|        |            |                   |                 |            |             |             |             |       |         |           |
|  **Count** | **Job-OZRCZA** | **\main\corerun.exe** |            **None** |     **100000** |  **5,207.5 μs** |    **82.89 μs** |    **77.54 μs** |  **1.00** |    **0.00** |         **-** |
|  Count | Job-PJSYDW |   \pr\corerun.exe |            None |     100000 |  6,424.4 μs |    89.90 μs |    84.10 μs |  1.23 |    0.03 |         - |
|        |            |                   |                 |            |             |             |             |       |         |           |
|  **Count** | **Job-OZRCZA** | **\main\corerun.exe** |            **None** |    **1000000** | **54,320.9 μs** | **1,079.00 μs** | **1,802.77 μs** |  **1.00** |    **0.00** |         **-** |
|  Count | Job-PJSYDW |   \pr\corerun.exe |            None |    1000000 | 63,753.0 μs |   984.47 μs |   822.07 μs |  1.22 |    0.03 |         - |
|        |            |                   |                 |            |             |             |             |       |         |           |
|  **Count** | **Job-OZRCZA** | **\main\corerun.exe** |        **Compiled** |      **10000** |    **296.5 μs** |     **1.90 μs** |     **1.69 μs** |  **1.00** |    **0.00** |         **-** |
|  Count | Job-PJSYDW |   \pr\corerun.exe |        Compiled |      10000 |    381.2 μs |     2.84 μs |     2.52 μs |  1.29 |    0.01 |         - |
|        |            |                   |                 |            |             |             |             |       |         |           |
|  **Count** | **Job-OZRCZA** | **\main\corerun.exe** |        **Compiled** |     **100000** |  **2,922.3 μs** |    **38.81 μs** |    **36.31 μs** |  **1.00** |    **0.00** |         **-** |
|  Count | Job-PJSYDW |   \pr\corerun.exe |        Compiled |     100000 |  3,804.2 μs |    26.95 μs |    23.89 μs |  1.30 |    0.02 |         - |
|        |            |                   |                 |            |             |             |             |       |         |           |
|  **Count** | **Job-OZRCZA** | **\main\corerun.exe** |        **Compiled** |    **1000000** | **29,236.7 μs** |   **311.23 μs** |   **291.13 μs** |  **1.00** |    **0.00** |         **-** |
|  Count | Job-PJSYDW |   \pr\corerun.exe |        Compiled |    1000000 | 38,461.8 μs |   316.45 μs |   280.53 μs |  1.32 |    0.02 |         - |
|        |            |                   |                 |            |             |             |             |       |         |           |
|  **Count** | **Job-OZRCZA** | **\main\corerun.exe** | **NonBacktracking** |      **10000** |    **835.2 μs** |     **8.26 μs** |     **7.32 μs** |  **1.00** |    **0.00** |         **-** |
|  Count | Job-PJSYDW |   \pr\corerun.exe | NonBacktracking |      10000 |    928.0 μs |     7.03 μs |     6.57 μs |  1.11 |    0.02 |         - |
|        |            |                   |                 |            |             |             |             |       |         |           |
|  **Count** | **Job-OZRCZA** | **\main\corerun.exe** | **NonBacktracking** |     **100000** |  **8,208.5 μs** |    **96.31 μs** |    **80.43 μs** |  **1.00** |    **0.00** |         **-** |
|  Count | Job-PJSYDW |   \pr\corerun.exe | NonBacktracking |     100000 |  9,362.0 μs |   146.36 μs |   122.22 μs |  1.14 |    0.02 |         - |
|        |            |                   |                 |            |             |             |             |       |         |           |
|  **Count** | **Job-OZRCZA** | **\main\corerun.exe** | **NonBacktracking** |    **1000000** | **83,772.2 μs** | **1,098.40 μs** | **1,027.44 μs** |  **1.00** |    **0.00** |         **-** |
|  Count | Job-PJSYDW |   \pr\corerun.exe | NonBacktracking |    1000000 | 93,933.0 μs |   999.32 μs |   834.48 μs |  1.12 |    0.02 |         - |

Reply:
Thanks, @joperezr.  A 30% regression, even worst-case, is a big pill to swallow.

I suggest you remove the change from this PR.  I'm completely on board with the goal of getting rid of that delegate-based implementation, but we need to first drive down the overheads of the other code paths.  I'd added the delegate-based path in .NET 5 because there was so much overhead associated with getting from the public API into the meat of the matching, and the delegate approach enabled bypassing most of that by keeping things in a tight loop around what's now Scan.  If we could drive down the worst-case regression from 30% to, say, 10%, I think it'd be worth making the switch and getting rid of the duplication/complication.

Reply:
Yup agreed, already pushed the change that undos the change in Count. I'll address the rest of your feedback tomorrow 😊